[SPARK-13148] [YARN] document zero-keytab Oozie application launch; add diagnostics

steveloughran opened the pull request:
This patch provides detail on what to do for keytabless Oozie launches of spark apps, and adds some debug-level diagnostics of what credentials have been submitted

steveloughran committed:
[SPARK-13148] support zero-keytab Oozie application launch on a secure cluster

SparkQA commented:
**[Test build #50599 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50599/consoleFull)** for PR 11033 at commit [`d7662b5`](https://github.com/apache/spark/commit/d7662b5c7b36b74a990366bee2489888353a6183).

SparkQA commented:
**[Test build #50599 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50599/consoleFull)** for PR 11033 at commit [`d7662b5`](https://github.com/apache/spark/commit/d7662b5c7b36b74a990366bee2489888353a6183).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50599/
Test FAILed.

steveloughran commented:
This now works: I've tested it by creating tokens, saving them to a file. pointing to in the env var then using spark-submit to bring up a cluster.

For the curious, you need 2-3 tokens: HDFS, RM, and, if Yarn timeline server is enabled, a timeline server delegation token.

```
Kind: HDFS_DELEGATION_TOKEN, Service: 192.168.56.101:8020, Ident: (HDFS_DELEGATION_TOKEN token 263 for stevel); HDFS_DELEGATION_TOKEN token 263 for stevel Renewer: yarn Issued: 03/02/16 13:26 Max Date: 10/02/16 13:26
Kind: RM_DELEGATION_TOKEN, Service: 192.168.56.101:8050, Ident: (owner=stevel@COTHAM, renewer=yarn, realUser=, issueDate=1454505993370, maxDate=1455110793370, sequenceNumber=5, masterKeyId=172); owner=stevel@COTHAM, renewer=yarn, realUser=, issueDate=1454505993370, maxDate=1455110793370, sequenceNumber=5, masterKeyId=172 Renewer: yarn Issued: 03/02/16 13:26 Max Date: 10/02/16 13:26
Kind: TIMELINE_DELEGATION_TOKEN, Service: 192.168.56.101:8188, Ident: (owner=stevel, renewer=yarn, realUser=, issueDate=1454505993925, maxDate=1455110793925, sequenceNumber=4, masterKeyId=2); owner=stevel, renewer=yarn, realUser=, issueDate=1454505993925, maxDate=1455110793925, sequenceNumber=4, masterKeyId=2 Renewer: yarn Issued: 03/02/16 13:26 Max Date: 10/02/16 13:26
```

steveloughran committed:
[SPARK-13148] address scalastyle issues

SparkQA commented:
**[Test build #50661 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50661/consoleFull)** for PR 11033 at commit [`e5e33c9`](https://github.com/apache/spark/commit/e5e33c98f2b518fc8088624dfd27e699197a244f).

SparkQA commented:
**[Test build #50661 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50661/consoleFull)** for PR 11033 at commit [`e5e33c9`](https://github.com/apache/spark/commit/e5e33c98f2b518fc8088624dfd27e699197a244f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50661/
Test FAILed.

steveloughran committed:
[SPARK-13148] fix NPE in log operation in test

SparkQA commented:
**[Test build #50685 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50685/consoleFull)** for PR 11033 at commit [`7f20dfb`](https://github.com/apache/spark/commit/7f20dfbc952bda03603db3d5af7972fad5016e45).

SparkQA commented:
**[Test build #50685 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50685/consoleFull)** for PR 11033 at commit [`7f20dfb`](https://github.com/apache/spark/commit/7f20dfbc952bda03603db3d5af7972fad5016e45).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50685/
Test PASSed.

tgravescs commented:
@steveloughran we launch spark via oozie in secure Hadoop environment without keytabs all the time.  Perhaps the heading on this isn't correct?

steveloughran commented:
This is to handle the launch situation
* secure cluster
* Oozie acquires tokens, sets to env var {{HADOOP_TOKEN_FILE_LOCATION}}
* launches client.

I guess if it is working for you, then yes, the user launching the code must be able to act as a proxy; this patch picks up the env var option instead.

I've changed the title and will do the same on the JIRA

tgravescs commented:
Sorry I'm still not clear on what you are trying to accomplish with this.  Why do we need to use the HADOOP_TOKEN_FILE_LOCATION?  

oozie (running with proxy user) setup for a secure cluster works just fine with Spark. It gets hdfs tokens for you and you can also tell oozie to get things like hive tokens for you. so oozie handles getting all the necessary credentials.  Perhaps there are other tokens oozie won't get for you?

steveloughran commented:
Yes: oozie gets all the tokens, but it has to hand them down. That env var is how it does it. As it is the spark client always tries to retrieve those tokens if UGI.isSecurityEnabled() == true

tgravescs commented:
oozie does it through the UGI and doAs and filling in the credentials.  Otherwise it wouldn't work for me now with Spark.  
That is why I'm trying to understand why this is needed.  I have no problems with it if there is a valid use case but I don't want extra security related code for nothing.  Perhaps you have different setup?  We run everything secure, oozie as proxy user.  Are you using some hybrid approach or not using proxy user.  Can you report the error you see?

You can prevent the spark client from trying to get hive and hbase tokens by turning them off.  You do have to do this as the client does try to always get them if its configured for those which will break since it tries from the oozie launcher running on the cluster.
config spark.yarn.security.tokens.${service}.enabled.  And you tell oozie to get the tokens for you via oozie workflow credentials commands.

steveloughran commented:
I know that you can skip credential pickup for hbase & hive, this patch also skips trying to collect any HDFS tokens for NNs, though of course that could be handled by setting "spark.yarn.access.namenodes"

What this does do is make the fact that env-supplied credentials are coming in more explicit, and dumping out their values.

I do now agree, by turning the right config options you could get away without being aware of the fact that the client is running in Oozie. In which case I propose a leaner patch which

1. Still does a dump of those tokens passed in and when they expire. This helps diagnose that problem where not enough tokens have been collected, and the problem only surfacing later.
1. Document how to use spark under oozie

what do you think?

tgravescs commented:
That sounds good.

steveloughran committed:
SPARK-13148 add security/oozie section in running-on-yarn.md

steveloughran committed:
SPARK-13148 extend running-on-yarn.md; switch Client back to master code with only a dump @ debug of credentials; this is mentioned in the docs

steveloughran commented:
Thomas: here's things trimmed back to do nothing but dump the credentials, and the docs updated to cover what needs to be done, and how to troubleshoot it in: [running-on-yarn.md](https://github.com/steveloughran/spark/blob/stevel/feature/SPARK-13148-oozie/docs/running-on-yarn.md)

I haven't done a test of the code; what needs to be reviewed in this iteration is really the docs.

SparkQA commented:
**[Test build #51561 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51561/consoleFull)** for PR 11033 at commit [`67efcfa`](https://github.com/apache/spark/commit/67efcfac585542f50eddbf9630cfdff357c1ad92).

tgravescs commented on docs/running-on-yarn.md:
```diff
@@ -441,3 +441,81 @@ If you need a reference to the proper location to put log files in the YARN so t
+# Running in a secure YARN cluster
+
+As covered in [security](security.html), Kerberos is used in a secure YARN cluster to
+authenticate principals with services —and so access these services and
```
typo... "and to"

steveloughran replied:
not really, I'm trying to say authenticate and then you can access the services. But I'll rewrite the whole para to make it clearer.

tgravescs commented on docs/running-on-yarn.md:
````diff
@@ -441,3 +441,81 @@ If you need a reference to the proper location to put log files in the YARN so t
+spark.yarn.access.namenodes hdfs://ireland.emea.example.org:8020/,hdfs://frankfurt.emea.example.org:8020/
+```
+
+Hadoop tokens expire. They can be renewed "for a while"; the Spark Application Master will automatically
````
The application master doesn't renew them, the yarn resource manager does.  Except for when keytabs are specified.

steveloughran replied:
I'm not sure what goes on there. I do know that in YARN-4653, I was told [the RM doesn't renew things](https://issues.apache.org/jira/browse/YARN-4653?focusedCommentId=15136094&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15136094), which is what I've written up in [Yarn Application Security](https://github.com/apache/hadoop/blob/trunk/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-site/src/site/markdown/YarnApplicationSecurity.md)

Maybe I'll just say "eventually, Hadoop tokens expire". That is true, and avoids the "who renews what" issue. And isn't relevant here except for the fact that you will need to play with keytabs for long-lived apps, which isn't the main point

tgravescs replied:
I believe he was mistaken. He was probably thinking of the other feature where YARN, in like 2.7 or 2.8, will actually get you an hdfs token for log aggregation.  The RM will renew tokens submitted as long as they have a renewer associated.  That is hdfs, hive, hbase, MR history, etc.

But I'm good with what you propose and just get rid of who does it.

steveloughran replied:
done

tgravescs commented on docs/running-on-yarn.md:
```diff
@@ -441,3 +441,81 @@ If you need a reference to the proper location to put log files in the YARN so t
+launched without a keytab, the responsibility for setting up security must be handed
+over to Oozie.
+ 
+The Oozie workflow configuration must be set up for Oozie to request all tokens, including:
```
it might be nice either to mention the oozie credentials parameter or just say see oozie documentation on how to do this

steveloughran replied:
referring to the oozie docs will probably be the most maintained. Which URL would you recommend?

tgravescs replied:
I guess probably just point to http://oozie.apache.org/ and say look at the documentation for your version and look for the authentication section.

SparkQA commented:
**[Test build #51561 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51561/consoleFull)** for PR 11033 at commit [`67efcfa`](https://github.com/apache/spark/commit/67efcfac585542f50eddbf9630cfdff357c1ad92).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51561/
Test PASSed.

steveloughran committed:
SPARK-13148: update the documentation with a link to the oozie site, but without any ref to a specific release's auth pages

SparkQA commented:
**[Test build #52056 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52056/consoleFull)** for PR 11033 at commit [`74cba66`](https://github.com/apache/spark/commit/74cba662724f7a67f973e24b912273cd92f43f58).

SparkQA commented:
**[Test build #52056 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52056/consoleFull)** for PR 11033 at commit [`74cba66`](https://github.com/apache/spark/commit/74cba662724f7a67f973e24b912273cd92f43f58).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52056/
Test PASSed.

steveloughran committed:
SPARK-13418 review and revise documentation

SparkQA commented:
**[Test build #52749 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52749/consoleFull)** for PR 11033 at commit [`3dbcac0`](https://github.com/apache/spark/commit/3dbcac0f6960c6707edad355d5f134074f8ab122).

SparkQA commented:
**[Test build #52749 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52749/consoleFull)** for PR 11033 at commit [`3dbcac0`](https://github.com/apache/spark/commit/3dbcac0f6960c6707edad355d5f134074f8ab122).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52749/
Test PASSed.

tgravescs commented on docs/running-on-yarn.md:
```diff
@@ -441,3 +441,91 @@ If you need a reference to the proper location to put log files in the YARN so t
+credentials for a job can be found on the [Oozie web site](http://oozie.apache.org/)
+in the "Authentication" section of the specific release's documentation.
+ 
+For Spark applications, the Oozie workflow must be set up for Oozie to request all tokens, including:
```
This isn't completely true as if I'm not doing hive or hbase I don't need those.  Can we change to clarify that those are optional.

steveloughran replied:
I wrote down the full details on how they are loaded, so nobody needs to look at the source to understand the conditions.

tgravescs replied:
sorry not sure I follow, I don't need to tell oozie to get hive and hbase tokens if I'm not going to use hive or hbase in my spark job.  So I'll I'm saying it perhaps like YARN timeline server below add in something that says if your application uses them.

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala:
```diff
@@ -326,6 +330,65 @@ class YarnSparkHadoopUtil extends SparkHadoopUtil {
+   * all tokens created in advance.
+   * @return the path to a credentials file if that is where the credentials came from
+   */
+  def environmentCredentialsFile() : Option[File] = {
```
did you mean to add this back in?

steveloughran replied:
no, looks like I got something wrong during some cherry picking

steveloughran committed:
[SPARK-13148]: details on hive and hbase tokens, rm method that had snuck in during rebase fixup

SparkQA commented:
**[Test build #52762 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52762/consoleFull)** for PR 11033 at commit [`f893af0`](https://github.com/apache/spark/commit/f893af00612f0bfc4fb149c80bfd739a36124633).

vanzin commented:
So this change is mostly documentation now, right? The title / summary look way out of date given that.

SparkQA commented:
**[Test build #52762 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52762/consoleFull)** for PR 11033 at commit [`f893af0`](https://github.com/apache/spark/commit/f893af00612f0bfc4fb149c80bfd739a36124633).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52762/
Test PASSed.

steveloughran commented:
1. I'll update
2. I think the extra credential dump should be pulled up into {{SparkHadoopUtil}}; it's not yarn-specific

steveloughran committed:
[SPARK-131458]
-Pull credential dump up to SparkHadoopUtil; It'd broader than just YARN.
-update docs to clarify HBase/Hive tokens are entirely optional.
-general review of text.

SparkQA commented:
**[Test build #53768 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53768/consoleFull)** for PR 11033 at commit [`9a37d62`](https://github.com/apache/spark/commit/9a37d629133d0743d476019893d886de80dd6d60).

tgravescs commented on docs/running-on-yarn.md:
```diff
@@ -452,3 +452,104 @@ If you need a reference to the proper location to put log files in the YARN so t
+to the authenticated principals.
+
+Hadoop services issue *hadoop tokens* to grant access to the services and data,
+tokens which the client must supply over Hadoop IPC and REST/Web APIs as proof of access rights.
```
These sentences sound a bit weird to me.  Perhaps something just like:

Hadoop services issue tokens to grant access to the services and data.  Clients must first acquire tokens for the services they will access and pass them along with their application.

tgravescs commented on docs/running-on-yarn.md:
````diff
@@ -452,3 +452,104 @@ If you need a reference to the proper location to put log files in the YARN so t
+spark.yarn.access.namenodes hdfs://ireland.example.org:8020/,hdfs://frankfurt.example.org:8020/
+```
+
+Hadoop tokens expire. They can be renewed "for a while".
````
I think we should explain this a bit more.  Either add more details about hard expiration and the renew period or just leave out the renew part.

steveloughran replied:
adding more details is something I'm scared of on the basis that I'd only repeat my own misunderstandings. 

What would you suggest?

steveloughran replied:
...I'll leave out the renew part.

tgravescs replied:
ok, sorry I missed your question go by.

SparkQA commented:
**[Test build #53768 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53768/consoleFull)** for PR 11033 at commit [`9a37d62`](https://github.com/apache/spark/commit/9a37d629133d0743d476019893d886de80dd6d60).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53768/
Test PASSed.

SparkQA commented:
**[Test build #57008 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57008/consoleFull)** for PR 11033 at commit [`4707ec6`](https://github.com/apache/spark/commit/4707ec624bf94cd6b2e592fc76640fafb8b137a9).

SparkQA commented:
**[Test build #57008 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57008/consoleFull)** for PR 11033 at commit [`4707ec6`](https://github.com/apache/spark/commit/4707ec624bf94cd6b2e592fc76640fafb8b137a9).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57008/
Test FAILed.

steveloughran commented:
@tgravescs the logging bit of the patch is in sync with master/ . 

Is there anything else you want me to do regarding the documentation to get it into a committable state?

tgravescs commented:
I have the comment above about the wording on one of the sentences, otherwise I think its fine.

steveloughran committed:
SPARK-13148 more sentence tuning

SparkQA commented:
**[Test build #58699 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58699/consoleFull)** for PR 11033 at commit [`f14768a`](https://github.com/apache/spark/commit/f14768aa6fce320ba07f7146755a833d651d16ac).

SparkQA commented:
**[Test build #58699 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58699/consoleFull)** for PR 11033 at commit [`f14768a`](https://github.com/apache/spark/commit/f14768aa6fce320ba07f7146755a833d651d16ac).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58699/
Test FAILed.

steveloughran committed:
[SPARK-13148] scalastyle

SparkQA commented:
**[Test build #58775 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58775/consoleFull)** for PR 11033 at commit [`56571da`](https://github.com/apache/spark/commit/56571da16cc0e71fd68c07abba26476fa51f7ddf).

SparkQA commented:
**[Test build #58775 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58775/consoleFull)** for PR 11033 at commit [`56571da`](https://github.com/apache/spark/commit/56571da16cc0e71fd68c07abba26476fa51f7ddf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58775/
Test PASSed.

tgravescs commented:
+1

steveloughran commented:
thanks
